### ISSUE TITLE
Windows installer removing user configuration

### ISSUE THREAD

**Issue by popojargo** [apache/couchdb#1989]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

When upgrading CouchDB on Windows, the user configuration seems to be removed. Only the couchdb.log file remains + `/data` folder.

## Steps to Reproduce

1. Install CouchDB
2. Change binding address to 0.0.0.0
3. Uninstall CouchDB
4. Reinstall CouchDB
5. Binding address will be back at `0.0.0.0`

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

User's configuration/data should remains on uninstallation.

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB Version used: 2.3.1 (Happens with 2.X versions)
* Browser name and version: Chrome 
* Operating System and version: Windows 10 Pro 64bit

## Additional context

[TIP]:  # ( Add any other context about the problem here. )

**Comment by popojargo** [comment]
I'll move that over `apache/couchdb-glazier` since it's related to Windows build only.

**Comment by popojargo** [comment]
I know that with WIX, we can easily configure some files so that they remains when uninstalling.

See this S/O response for example: https://stackoverflow.com/a/2631261/5236185

I can submit a pull request for that but I would like to know what are the files to keep? 

The only one that I'm sure so far is: `/etc/local.ini`

Is there any other files?

**Comment by popojargo** [comment]
Updates from @wohali : etc/local.ini and the entire etc/local.d directory

**Comment by popojargo** [comment]
I've figured out how to keep the files after uninstalling but I don't know how to apply it to the current setup. We generate the files components using Heat. So each file has  GUID and I can't find how to modify them using Wix files without using XML transoformations when generating files.

**Comment by big-r81** [comment]
Hi,
I think there are only two options:

1. make a static fragment of the config files with Atrributes keeping these files during an upgrade and remove them from harvesting with a XML transformation or
2. use a XML transformation and modify the Attributes of the config file entries during harvesting from the heat-tool.

See: https://stackoverflow.com/questions/8034798/wix-installer-using-xslt-with-heat-exe-to-update-attributes

**Comment by popojargo** [comment]
@big-r81  Got any experience with wix and heat? I've tried something like this: https://github.com/apache/couchdb-glazier/pull/10 but it doesn't fix the actual problem: the upgrade.

**Comment by wohali** [comment]
This is now fixed in the new Windows CouchDB 3.0 installer (to be uploaded with the 3.0 release), as part of changes necessary to support prompting for the admin user/password on install.

The new installer now takes these actions, in order:
* prompts for the mandatory admin user & password
* Installs the user-editable config files as `vm.args.dist` and `local.ini.dist`
* Creates the file `etc\local.d\10-admins.ini` with the admin user specified in the installer, but **only if** the file does not yet exist
* Copies the `vm.args.dist` file to `vm.args`, but **only if** that file doesn't exist yet
* Copies the `local.ini.dist` file to `local.ini`, but **only if** that file doesn't exist yet
* Starts the CouchDB service, which encrypts the plaintext password in `etc\local.d\10-admins.ini` immediately

On uninstall, the new uninstaller:
* correctly stops both the Apache CouchDB service and the `epmd.exe` process
* only removes the `vm.args` file, but **only if** it exactly matches the `vm.args.dist` file
* only removes the `local.ini` file, but **only if** it exactly matches the `local.ini.dist` file
* removes the `vm.args.dist` and `local.ini.dist` files along with the rest of the install
* leaves behind the `etc\local.d\10-admins.ini` file, along with the contents of `data\`

The new behaviour above is supported through a WiX custom action written in C#.

Upgrades for now are still uninstall-then-reinstall, but the above safeguards mean that both your data and the user configurations are untouched during the process.

@popojargo Keep in mind that when you uninstall CouchDB 2.3.1, it'll still remove the user config files, but after you install 3.0.0, it won't happen again.

One final note, since people may want to know: if you want to do a silent install of CouchDB,  here's how:

```
msiexec /i apache-couchdb-3.0.0.msi /quiet ADMINUSER=admin ADMINPASSWORD=hunter2 /norestart
```

The uninstall is similar:

```
msiexec /x apache-couchdb-3.0.0.msi /quiet /norestart
```
